mds: move inline data to data pool object

mchangir opened the pull request:
If inline data version is not CEPH_INLINE_NONE then move data to data
pool object and set inline data version to CEPH_INLINE_NONE.

Fixes: https://tracker.ceph.com/issues/52916
Signed-off-by: Milind Changire <mchangir@redhat.com>





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

lxbsz commented on src/mds/CInode.h:
```diff
@@ -1050,6 +1050,15 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
   static void dump_validation_results(const validated_data& results,
                                       ceph::Formatter *f);
+  bool has_inline_data() {
+    if (is_normal() and is_file()) {
```
`s/and/&&/` ?

mchangir replied:
replaced `and` with `&&`

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -12955,6 +12958,161 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+
+  dout(20) << "(uninline_data) mutation "
+	   << (onfinish.wait() == 0 ? "succeeded" : "failed") << dendl;
+}
```
Maybe we should journal this. And do the uninline in the auth MDS only and notify all the replica MDSes at the same time ?

lxbsz commented:
There have at least 2 bugs for now in the inline code in kernel [1] and [2], which for [1] it's hard to fix, once this feature is support, IMO we'd better to make the files have inline data to be readonly, and at the same time add one option, which the default is false, to enable readwrite if needed by users.

[1] https://patchwork.kernel.org/project/ceph-devel/patch/20220506143052.479799-1-xiubli@redhat.com/
[2] https://tracker.ceph.com/issues/52921

lxbsz commented on src/mds/CInode.h:
```diff
@@ -1050,6 +1050,15 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
                                       ceph::Formatter *f);
+  bool has_inline_data() {
+    if (is_normal() && is_file()) {
+      auto mi = get_inode();
```
Shouldn't we use the projected inode here ?

mchangir replied:
done

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -12969,6 +12972,167 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+  }
+};
+
+struct C_IO_DataUninlined : public Context {
```
You should inherit from `MDSIOContext` instead here, the `Context::complete()` won't help hold the `mds_lock`.

mchangir replied:
I tried using MDSIOContext, but that led to deadlocks.
Still trying to understand the various flavours of contexts.

lxbsz replied:
I see you are also doing the `mds_lock` in the `complete()` in `C_IO_DataUninlined`, if use the `MDSIOContext`, then you should remove the `mds_lock` here.

mchangir replied:
done

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -12969,6 +12972,167 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+      mdr->mark_event(event_str);
+
+      auto fin = new C_MDC_DataUninlinedSubmitted(mdr);
+      mdlog->submit_entry(le, fin);
```
IMO you should journal the request in `uninline_data_work()` instead just before you do anything as early as possible.

mchangir replied:
any other sequence of journaling (which requires mds_lock) causes crashes if not done atomically under the mds_lock

lxbsz replied:
When the `uninline_data_work()` is called the `mds_lock` has already been hold. For this you could refer to `Server::handle_client_setattr()` just follow what the inode truncate does.

mchangir replied:
`uninline_data_work()` is an asynchronous activity. At the end of the function, a continuation context is passed to the objecter trying to mutate the file. When the objecter finishes its job, the continuation is called to notify work completion and then the event is recorded in the journal.

I tried to start logging the event in `uninline_data_work()` as you suggested, but the mds crashed with a traceback when trying to start another entry ... this is because although `uninline_data_work()` had acquired the `mds_lock`, further work was delegated to an asynchronous objecter activity which caused the `mds_lock` to be released since the `scoped_lock` lost its scope. Do you recommend locking the `mds_lock` in `uninline_data_work()` and unlock in `C_IO_DataUninlined::finish()` ? I ran a test with these changes, but neither the MDS crashed nor were the files uninlined.

The inode truncate workflow in `Server::handle_client_setattr()` is a bit different since all the data IO activity has been taken care of by the client by interacting directly with the OSDs and has only delegated updating attr metadata to the MDS. In case of `uninline_data_work()` the file data IO needs to be taken care of by the MDS as well followed by metadata update and logging to the journal.

mchangir replied:
wouldn't writing data blocks later be a problem is there's insufficient space to write out the data blocks after journaling the changes apriori ?

on the other hand, if the journal writing fails, then clients can continue to read from the inode as usual even if the blocks have been written to the 0th file object in the data pool

lxbsz replied:
> wouldn't writing data blocks later be a problem if there's insufficient space to write out the data blocks after journaling the changes apriori ?

You should check it before journaling.

> 
> on the other hand, if the journal writing fails, then clients can continue to read from the inode as usual even if the blocks have been written to the 0th file object in the data pool

We shouldn't be care about the journal failing case, or the other normal cases will have the same issue, and how do they handle it ?

Thinking more about it, this is a little special. Because it not only will handle the metadatas but also the datas of a file. If we journal it and then set the `version` to `CEPH_INLINE_NONE` and then return success, then when a client want to read it and if the journals are still not expired yet. The read may fail or just return 0, because only after the journals expired will the `0th` object created and have the inline data.

So this should be okay, we shouldn't journal it.

mchangir commented:
[These set of teuthology jobs](http://pulpito.ceph.com/mchangir-2022-06-07_05:18:32-fs-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/) ran the data uninlining and scrubbing over a set of ~8000 files and uninlined 5145 files in total.

lxbsz commented:
> > There have at least 2 bugs for now in the inline code in kernel [1] and [2], which for [1] it's hard to fix, once this feature is support, IMO we'd better to make the files have inline data to be readonly, and at the same time add one option, which the default is false, to enable readwrite if needed by users.
> 
> How is supporting this feature (un-inline during scrub) effecting the bugs you mention? The bugs exist irrespective of this feature availability? Maybe I'm, misunderstanding something? @lxbsz
> 
> > [1] https://patchwork.kernel.org/project/ceph-devel/patch/20220506143052.479799-1-xiubli@redhat.com/ [2] https://tracker.ceph.com/issues/52921

There have 2 known bugs in kclient and one of them is hard to be fixed and another one we won't fix it since we are planing to disable the inline data feature in late future.

As we discussed in kernel mail list, since the inline_data feature is half implemented and we were planing to disable it and finally will drop the corresponding code in kclient, and was waiting this feature.

To support that we need to make the files to be readonly from MDS side if they have inlined data after this PR gets merged and then request to uninline them by using the scrub tool first and then continue. I have discussed this with @mchangir and I can add one following patch to do it after this.

batrick commented on src/mds/MDCache.cc:
```diff
@@ -12969,6 +12972,161 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+  auto& snapc_ref = (snaprealm ? snaprealm->get_snap_context() : snapc);
+
+  ObjectOperation create_ops;
+  create_ops.create(false);
```
The file's first data object may or may not exist. If it does not, we probably should write out the backtrace as well? This needs to sync with CInode. Perhaps some refactoring?

batrick replied:
It may not if the journal has not expired the segment for a new inode (IIRC).

lxbsz replied:
Test this and @batrick is correct.

lxbsz replied:
I don't think you need to do this here.  If the `0th` object is not created yet that work should be already queued in the journals. And when those journals expired and commited they should update the backtraces for you.

Or in `Client.cc`  and kclient we should do it too when uninlining the inlinedata.

mchangir commented on src/mds/MDSRank.h:
```diff
@@ -634,6 +639,8 @@ class MDSRank {
     mono_time starttime = mono_clock::zero();
     boost::asio::io_context& ioc;
+
+    ceph_tid_t internal_reqid = 1;
```
hadn't looked closely at `make_internal_request<>()` implementation back then, which assigns a request id
will remove this line

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -152,6 +153,19 @@ class C_RetryScrub : public MDSInternalContext {
   MDSCacheObject *obj;
 };
 
+class C_ScrubFileInode : public MDSInternalContext {
```
will remove this unwanted continuation class

mchangir commented on src/mds/mdstypes.h:
```diff
@@ -856,6 +856,8 @@ void inode_t<Allocator>::dump(ceph::Formatter *f) const
   f->dump_int("export_pin", export_pin);
   f->dump_int("export_ephemeral_random_pin", export_ephemeral_random_pin);
   f->dump_bool("export_ephemeral_distributed_pin", export_ephemeral_distributed_pin);
+  f->dump_unsigned("inline_data_version", inline_data.version);
```
created a separate commit

mchangir commented on src/mds/CDir.h:
```diff
@@ -460,6 +460,7 @@ class CDir : public MDSCacheObject, public Counter<CDir> {
   bool is_exporting() { return state & STATE_EXPORTING; }
   bool is_importing() { return state & STATE_IMPORTING; }
   bool is_dirty_dft() { return state & STATE_DIRTYDFT; }
+  bool is_stable() { return (state == STATE_COMPLETE); }
```
removed this line

vshankar commented on qa/suites/fs/functional/tasks/uninlining.yaml:
```diff
@@ -0,0 +1,25 @@
+        debug client: 20
+      mds:
+        # to force replication without waiting for hit ratio ot ramp up
+        mds_bal_replicate_threshold: 1
```
What I meant was to include replication via a yaml for overall fs suite tests.

mchangir commented on src/mds/MDCache.cc:
```diff
@@ -12969,6 +12972,161 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+  public:
+  C_IO_DataUninlined(MDRequestRef r, MDSRank *mds) : MDSIOContext(mds), mdr(r) {}
+
+  virtual void print(std::ostream& os) const {
```
`MDSIOContext` has a pure virtual function with the same signature ... so the `print()` function becomes mandatory
I'll find something to print just for conformance.

batrick commented on qa/tasks/cephfs/test_uninlining.py:
```diff
@@ -0,0 +1,141 @@
+                                       json_out['inline_data_version']))
+        return info
+
+    def test_data_uninlining(self):
```
> So far, as far as an inode is concerned, I've found that the only damage is missing backtrace on on-disk object

That may be the case. If so, ignore this request to damage metadata.

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -1129,3 +1144,19 @@ void ScrubStack::handle_mds_failure(mds_rank_t mds)
+{
+  dout(20) << "(uninline_data) starting data uninlining for " << *in << dendl;
+
+  MDRequestRef mdr = in->mdcache->request_start_internal(CEPH_MDS_OP_UNINLINE_DATA);
```
done

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -1129,3 +1144,19 @@ void ScrubStack::handle_mds_failure(mds_rank_t mds)
+  mdr->internal_op_private = in;
+  mdr->internal_op_finish = fin;
+
+  in->mdcache->uninline_data_work(mdr);
```
done

mchangir commented:
rebased to latest

batrick commented on src/mds/ScrubHeader.h:
```diff
@@ -19,14 +19,17 @@
 #include <memory>
 #include <string>
 #include <string_view>
+#include <condition_variable>
```
This line should not be necessary. `common/fair_mutex.h` picks that up from `common/ceph_mutex.h`.

I'm not sure why, even so, you get this:

```
FAILED: src/crimson/CMakeFiles/crimson-common.dir/__/msg/Message.cc.o 
/usr/bin/ccache /usr/bin/clang++-12  -DBOOST_ALL_NO_LIB -DBOOST_ASIO_DISABLE_CONCEPTS -DBOOST_ASIO_DISABLE_THREAD_KEYWORD_EXTENSION -DBOOST_ASIO_USE_TS_EXECUTOR_AS_DEFAULT -DBOOST_IOSTREAMS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_RANDOM_DYN_LINK -DBOOST_THREAD_DYN_LINK -DCEPH_INSTALL_DATADIR=\"/usr/local/share/ceph\" -DCEPH_INSTALL_FULL_PKGLIBDIR=\"/usr/local/lib/ceph\" -DCMAKE_INSTALL_LIBDIR=\"lib\" -DHAVE_CONFIG_H -DSEASTAR_API_LEVEL=6 -DSEASTAR_DEBUG -DSEASTAR_DEBUG_SHARED_PTR -DSEASTAR_DEFAULT_ALLOCATOR -DSEASTAR_SCHEDULING_GROUPS_COUNT=16 -DSEASTAR_SHUFFLE_TASK_QUEUE -DSEASTAR_TYPE_ERASE_MORE -DWITH_SEASTAR=1 -D_FILE_OFFSET_BITS=64 -D_GNU_SOURCE -D_REENTRANT -D_THREAD_SAFE -D__CEPH__ -D__STDC_FORMAT_MACROS -D__linux__ -Isrc/include -I../src -I../src/seastar/include -Isrc/seastar/gen/include -isystem /opt/ceph/include -isystem include -isystem ../src/xxHash -isystem ../src/rapidjson/include -isystem ../src/jaegertracing/opentelemetry-cpp/api/include -isystem ../src/jaegertracing/opentelemetry-cpp/exporters/jaeger/include -isystem ../src/jaegertracing/opentelemetry-cpp/ext/include -isystem ../src/jaegertracing/opentelemetry-cpp/sdk/include -g -Werror -fPIC   -fno-builtin-malloc -fno-builtin-calloc -fno-builtin-realloc -fno-builtin-free -Wall -fno-strict-aliasing -fsigned-char -Wtype-limits -Wignored-qualifiers -Wpointer-arith -Werror=format-security -Winit-self -Wno-unknown-pragmas -Wnon-virtual-dtor -Wno-ignored-qualifiers -ftemplate-depth-1024 -Wpessimizing-move -Wredundant-move -Wno-inconsistent-missing-override -Wno-mismatched-tags -Wno-unused-private-field -Wno-address-of-packed-member -Wno-unused-function -Wno-unused-local-typedef -Wno-varargs -Wno-gnu-designator -Wno-missing-braces -Wno-parentheses -Wno-deprecated-register -DCEPH_DEBUG_MUTEX -D_GLIBCXX_ASSERTIONS -fdiagnostics-color=auto -Wno-non-virtual-dtor -std=gnu++17 -U_FORTIFY_SOURCE -DSEASTAR_SSTRING -Wno-error=unused-result "-Wno-error=#warnings" -fstack-clash-protection -fsanitize=address -fsanitize=undefined -fno-sanitize=vptr -std=c++17 -MD -MT src/crimson/CMakeFiles/crimson-common.dir/__/msg/Message.cc.o -MF src/crimson/CMakeFiles/crimson-common.dir/__/msg/Message.cc.o.d -o src/crimson/CMakeFiles/crimson-common.dir/__/msg/Message.cc.o -c ../src/msg/Message.cc
In file included from ../src/msg/Message.cc:141:
In file included from ../src/messages/MMDSCacheRejoin.h:20:
In file included from ../src/mds/CInode.h:37:
In file included from ../src/mds/CDentry.h:34:
In file included from ../src/mds/ScrubHeader.h:25:
```

`common/ceph_mutex.h` should be defining `ceph::condition_variable` here:

https://github.com/ceph/ceph/blob/dab3364f179167df026128a0d51a1dbe52d1f820/src/common/ceph_mutex.h#L80

I think you'll need to look at the preprocessor output to debug this.

mchangir replied:
I can see that `-DWITH_SEASTAR=1` has been defined on the build command-line; and looking at `src/common/ceph_mutex.h`, it is obvious that `condition_variable` is not defined when `WITH_SEASTAR` is deifned

So, if Jenkins builds are now scheduled to always build with `WITH_SEASTAR=1`, then `ceph_mutex.h` needs to define `condition_variable` appropriately for the Seastar case.

batrick commented on src/mds/ScrubHeader.h:
```diff
@@ -74,6 +92,10 @@ class ScrubHeader {
   unsigned epoch_last_forwarded = 0;
   unsigned num_pending = 0;
+  // uninline failed info mutex
+  ceph::fair_mutex ufi_mutex;
```
I think you can avoid this problem by simply putting this `fair_mutex` in a container (Pimpl C++ pattern). Then including this header won't require knowing its definition because you'll include it in `ScrubHeader.cc`. You'll need to move the definition of these methods to `ScrubHeader.cc` too.

mchangir replied:
I've done away with the fair_mutex; its not required.

mchangir commented:
[Successful Teuthology Job](https://pulpito.ceph.com/mchangir-2022-07-20_09:11:34-fs-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/)

lxbsz commented on src/mds/CInode.h:
```diff
@@ -1050,6 +1050,15 @@ class CInode : public MDSCacheObject, public InodeStoreBase, public Counter<CIno
+  bool has_inline_data() {
+    if (is_normal() && is_file()) {
+      auto pin = get_projected_inode();
+      if (pin->inline_data.version != CEPH_INLINE_NONE) {
```
Maybe this is incorrect. In case the inlinedata is enabled then for all the new CInodes the default version is `1` here. Which has no inlinedata.

mchangir replied:
are you concerned that the attempt to `write()` to the 0th file object is a no-op when the inline data length is zero ?
do you want to avoid that round trip to the osd ?

lxbsz replied:
IMO, we should also skip uninlining the inline data if the `pin->inline_data.version == 1`. Actually there is no inline data for this inode.

lxbsz replied:
Or the uninline data work will iterate all the CInodes and do the fake uninline work, even there is no inline data in any file.

mchangir replied:
for `write`: so, now the write takes place conditionally only if inline_data length is a non-zero positive value
for `create`: if I completely remove the `create` op, then uninline ops fail with ENOENT; at least for newly created files

will refresh the PR once I've addressed comments

mchangir replied:
`write` has been made conditional only if inline data length is non-zero

lxbsz replied:
In `void MDCache::uninline_data_work(MDRequestRef mdr)` I see after checking the `if (!in->has_inline_data()) {` if the inlindata version is `1` it will continue do a lot more work after that. IMO we can just skip them ?

mchangir replied:
If we are planning on deprecating and eventually decommissioning **inline data** feature, then we'd need to do all the work that needs to be done to move the information to the 0th file object. Inline data version and the xattr does need to be projected and logged to the journal. I don't think we can skip this work even if **inline data length** is zero irrespective of the version ... since inline data may have been written within the inline data length bounds and truncated to zero length many times. If so, then the inline data version may not be 1 but the inline data length may be zero.

@batrick could you take a peek here and suggest appropriately

batrick replied:
So, I think there is a structural problem in this code: we shouldn't be looking at the projected version at all. If we're uninlining data then clients should not be writing any new inline data, no? I think before any automatic uninlining is done via scrub, the MDS should ensure there are no inline data writes in play. If there are, scrub should not do any uninlining and note a health warning that inline is still in use.

Also, I forget: how are you preventing new inline writes by clients?

batrick replied:
>  I presumed that acquiring xlocks causes cap revokes and/or commits pending writes, so that eventually acquiring the xlock in the uninline code path fences the client from causing any racy writes.

Sure (but verify with tests!).

> So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.

Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.

I take back what I said about not looking at the projected data. That won't be practical. We do however need to prevent clients from doing new inline writes.

mchangir replied:
> > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> 
> Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.

Wouldn't the client block for `write` (`Fw`) caps for the truncate op ?
Or could things turn ugly with an async truncate op ?

@lxbsz could you comment here about the client behavior in this case

lxbsz replied:
> > > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> > 
> > 
> > Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.
> 
> Wouldn't the client block for `write` (`Fw`) caps for the truncate op ? 

The `Fw` is not a must for the size truncating, please see https://github.com/ceph/ceph/blob/main/src/client/Client.cc#L7868-L7895.

If the `Fx` cap is not issued it will send a sync `setattr` request to MDS, or the clients could buffer the size change.

> Or could things turn ugly with an async truncate op ?
> 

Then in the MDS side, the `Server::handle_client_setattr()` will require the `xlock` for the file. So the truncate request should be queued to wait the scrub uninline work to finish.

I think this should be okay and I don't see a racy for this.

> @lxbsz could you comment here about the client behavior in this case

But to be safe, I suggest to make the current file to be readonly while uninlining the inline data.

> > So, I think that once an xlock has been granted to the uninline code path, the has_inline_data() function need not look at the projected inode.
> 
> Um, the client could still have dirtied data (like a truncate) which requires looking at the projected inode.
>

Won't the dirtied data be flushed back before the scrub uninline work successfully getting the `xlock` ? After that no clients could buffer the new change.

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -12980,6 +12983,165 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+
+  char oid_buf[32];
+  snprintf(oid_buf, sizeof(oid_buf), "%llx.00000000", (long long unsigned)ino().val);
+  object_t oid = oid_buf;
```
Why not use the `InodeStoreBase::get_object_name()` to get the `oid` ?

mchangir replied:
ack

mchangir commented on src/include/cephfs/types.h:
```diff
@@ -750,6 +750,7 @@ void inode_t<Allocator>::dump(ceph::Formatter *f) const
   f->dump_int("export_ephemeral_random_pin", export_ephemeral_random_pin);
   f->dump_bool("export_ephemeral_distributed_pin", export_ephemeral_distributed_pin);
 
+
```
deleted spurious empty line

mchangir commented:
@lxbsz 

for `write`: so, now the write will take place conditionally only if inline_data length is a non-zero positive value
for `create`: if I completely remove the `create` op, then uninline ops fail with ENOENT; at least for newly created files

will refresh the PR once I've addressed comments

mchangir commented:
[Successful Teuthology Job](http://pulpito.ceph.com/mchangir-2022-07-22_07:24:08-fs-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/)

lxbsz commented on src/mds/MDCache.cc:
```diff
@@ -12980,6 +12983,180 @@ void MDCache::enqueue_scrub_work(MDRequestRef& mdr)
+  object_t oid = InodeStoreBase::get_object_name(ino(), frag_t(), "");
+  SnapContext snapc;
+  SnapRealm *snaprealm = in->find_snaprealm();
+  auto& snapc_ref = (snaprealm ? snaprealm->get_snap_context() : snapc);
```
BTW, is the `snapc` really needed ? From the code the `find_snaprealm()` should always return a snaprealm, which shouldn't be `nullptr`.

lxbsz replied:
It seems I am wrong here. Please ignore it.

mchangir commented:
rebased code

gregsfortytwo commented:
Jenkins test this please

mchangir commented:
rebased code

lxbsz reviewed (approved):
LGTM.

vshankar commented:
@mchangir FYI - http://pulpito.front.sepia.ceph.com/vshankar-2022-08-25_07:08:30-fs-wip-vshankar-uns-20220825-092856-testing-default-smithi/

vshankar commented:
> @mchangir FYI - http://pulpito.front.sepia.ceph.com/vshankar-2022-08-25_07:08:30-fs-wip-vshankar-uns-20220825-092856-testing-default-smithi/

Most of the failures are due to the ongoing distro mess. I'll schedule a run with https://github.com/ceph/ceph/pull/47805

vshankar commented:
@mchangir https://pulpito.ceph.com/vshankar-2022-09-01_13:47:56-fs-wip-vshankar-uns-20220901-150020-testing-default-smithi/

chugging slowly....

mchangir commented:
@vshankar [new set of teuthology jobs](http://pulpito.front.sepia.ceph.com/mchangir-2022-09-12_07:34:04-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/)

mchangir commented:
@vshankar fyi
* [qa: timeout during unwinding of qa/workunits/suites/fsstress.sh](https://tracker.ceph.com/issues/57610)
* [qa: failure during qa/workunits/fs/snaps/snaptest-git-ceph.sh](https://tracker.ceph.com/issues/57611)
* [qa: segmentation fault during qa/workunits/libcephfs/test.sh](https://tracker.ceph.com/projects/cephfs/issues/new)

vshankar commented:
> @vshankar fyi
> 
> * [qa: timeout during unwinding of qa/workunits/suites/fsstress.sh](https://tracker.ceph.com/issues/57610)
> * [qa: failure during qa/workunits/fs/snaps/snaptest-git-ceph.sh](https://tracker.ceph.com/issues/57611)
> * [qa: segmentation fault during qa/workunits/libcephfs/test.sh](https://tracker.ceph.com/projects/cephfs/issues/new)

Could you rerun the failed tests. I can see one new failure, but that _might_ be a race test case.

vshankar commented:
> > @vshankar fyi
> > 
> > * [qa: timeout during unwinding of qa/workunits/suites/fsstress.sh](https://tracker.ceph.com/issues/57610)
> > * [qa: failure during qa/workunits/fs/snaps/snaptest-git-ceph.sh](https://tracker.ceph.com/issues/57611)
> > * [qa: segmentation fault during qa/workunits/libcephfs/test.sh](https://tracker.ceph.com/projects/cephfs/issues/new)
> 
> Could you rerun the failed tests. I can see one new failure, but that _might_ be a race test case.

http://pulpito.front.sepia.ceph.com:80/mchangir-2022-09-20_09:35:11-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/

One related failure that's reproducible -- http://pulpito.front.sepia.ceph.com/mchangir-2022-09-20_09:35:11-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/7038745

Rest are known failures. Let's get that fixed and this is good to merge! Nice work @mchangir

mchangir commented:
[Teuthology Jobs](https://pulpito.ceph.com/mchangir-2022-09-28_15:45:41-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/) tested with fix for `test_scrub_pause_and_resume_with_abort`
(note: the data uninlining job never ran in this set of jobs)

mchangir commented:
Another [set of jobs](https://pulpito.ceph.com/mchangir-2022-09-28_16:45:33-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/) exclusively running `tasks/scrub` to test `test_scrub_pause_and_resume_with_abort`

mchangir commented:
[Teuthology Jobs](https://pulpito.ceph.com/mchangir-2022-09-29_02:50:00-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/) after small correction to `test_scrub_pause_and_resume_with_abort` code

vshankar commented:
> [Teuthology Jobs](https://pulpito.ceph.com/mchangir-2022-09-29_02:50:00-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/) after small correction to `test_scrub_pause_and_resume_with_abort` code

Looks good. There are a couple of running jobs. Let's merge this once that finishes :)

vshankar commented:
jenkins retest this please

mchangir commented:
jenkins test make check arm64

mchangir commented:
jenkins test make check arm64

mchangir commented:
jenkins test make check arm64

vshankar commented:
@batrick Do you have any more comments on this change? I see there is one pending change request from you.

mchangir commented:
Last update contains:
* individual uninline failures are pushed to DamageTable
* individual scrub stats now available via scrub status command
* scrub stats can now be purged manually and automatically via timer event (defaulting to 1 day minimum)

mchangir commented:
@batrick this is almost ready

mchangir commented:
@vshankar this is almost ready; please take a look;

lxbsz commented:
@mchangir What's the status of this PR now ?

mchangir commented:
> @mchangir What's the status of this PR now ?

PR is ready for review (apart from merge conflicts if any)

lxbsz commented:
> > @mchangir What's the status of this PR now ?
> 
> PR is ready for review (apart from merge conflicts if any)

Sure. I will review it again today. After this we have more work to do the remove the inline code from both ceph and kclient.

lxbsz commented on qa/tasks/cephfs/test_uninlining.py:
```diff
@@ -0,0 +1,323 @@
+        # Extract test data tarball
+        # FIXME
+        log.debug("extracting tarball")
+        cmd = 'tar -x -z -f linux-5.4.tar.gz linux-5.4/fs/ceph linux-5.4/fs/orangefs linux-5.4/fs/ext2'
```
BTW, shouldn't we remove the `linux-5.4/` to make it clean for each test ? Doesn't it matter here ?

lxbsz replied:
Okay, please ignore this, I saw you have done this.

mchangir replied:
The `linux-5.4` dir is removed on line 261 after each iteration.

lxbsz commented on qa/tasks/cephfs/test_uninlining.py:
```diff
@@ -0,0 +1,323 @@
+
+        # Get configured max inline data size
+        log.debug("getting client_max_inline_size")
+        idsize = self.fs.fs_config.get('client_max_inline_size', 4096)
```
IMO we should strict the mounts to `ceph-fuse`. Because the kernel mount is buggy when the `client_max_inline_size` > 4KB. 

Or you could just force set `client_max_inline_size` to 4KB for this test ? Which could work for both kclient and fuse mounts.

lxbsz replied:
It's okay, just want to make sure in future someone won't touch this in this test.

And maybe we can just add some comments to give the users a hint if they hit issues between fuse and kernel clients for their existing inline data usage.

mchangir replied:
I'll add a comment in the python test code.

lxbsz commented on src/mds/ScrubStack.cc:
```diff
@@ -693,6 +693,32 @@ void ScrubStack::scrub_status(Formatter *f) {
+    f->open_object_section(tag);
+    {
+      f->dump_stream("start_time") << ctrs.start_time;
+      f->dump_string("path", (ctrs.origin_path == "" ? "/"s : ctrs.origin_path));
```
BTW, is the `s` suffix needed here ?

lxbsz commented on src/mds/ScrubStack.cc:
```diff
@@ -98,9 +98,51 @@ int ScrubStack::_enqueue(MDSCacheObject *obj, ScrubHeaderRef& header, bool top)
+void ScrubStack::purge_old_scrub_counters()
+{
+  // "mds_scrub_stats_review_period" must be in number of days
+  uint64_t mds_scrub_stats_review_period = g_conf().get_val<uint64_t>("mds_scrub_stats_review_period");
```
Maybe you could save this in a local member instead of reading this from the `g_conf()` per `tick()` ?

mchangir replied:
I'd thought of that too, but then didn't want to burden myself about how to handle changes to this config item at run time.

lxbsz replied:
There were several commits improved the code like this in `mds/client`.

mchangir replied:
is it okay if I add a new commit to handle this via a `handle_conf_change()` method ?

mchangir commented:
[Teuthology job](http://pulpito.front.sepia.ceph.com/mchangir-2023-02-06_11:39:42-fs-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/) with changes:
* rebase to upstream main
* comment in `test_uninlining.py` about `client_max_inline_size` (4096) limit for kclient
* new commit in `ScrubStack` to handle conf change to `mds_scrub_stats_review_period`

lxbsz commented:
LGTM. Thanks.

vshankar commented:
@mchangir Will pick this up in my next test run.

mchangir commented:
rebased

vshankar commented:
jenkins retest this please

vshankar commented:
https://pulpito.ceph.com/?branch=wip-vshankar-testing-20230421.044040

vshankar commented:
jenkins test make check arm64

mchangir commented:
rebased

mchangir commented:
rebased

mchangir commented:
jenkins test make check arm64

mchangir commented:
jenkins test make check

mchangir commented:
jenkins test windows

mchangir commented:
rebased

mchangir commented:
fixed build errors due to missing `mdlog->start_entry()` method

mchangir commented:
rebased

mchangir commented:
rebased

mchangir commented:
rebased

mchangir commented:
rebased

mchangir commented:
rebased

vshankar commented:
> rebased

@mchangir We should test this out asap given that we are looking at (new) trackers that are possible related to inline data being enabled. How does your test queue looks like for this week?

mchangir commented:
> > rebased
> 
> @mchangir We should test this out asap given that we are looking at (new) trackers that are possible related to inline data being enabled. How does your test queue looks like for this week?

just one pending 200+ job fs suite run to review when it gets back with the results

mchangir commented:
rebase and bug fixes

vshankar commented:
> @vshankar QA Status: [12 of 12 GREEN QA Runs](https://pulpito.ceph.com/mchangir-2024-04-01_14:26:20-fs:functional-wip-mchangir-testing1-main-20240328-072412-distro-default-smithi/)

Nice 👍

Do you plan to run this through the fs suite fully again? Esp after a bunch of qa related fixes get merged.

vshankar commented:
@mchangir I recall you ran this through fs suite. Any updates on the results?

mchangir commented:
> @mchangir I recall you ran this through fs suite. Any updates on the results?

I had mentioned about my plan to do so.
Awaiting build completion to start testing jobs.
Redmine Tracker: https://tracker.ceph.com/issues/65680

vshankar commented:
@mchangir Could you please link the tracker for the MDS memleak you created when validating run for this change?

mchangir commented:
> @mchangir Could you please link the tracker for the MDS memleak you created when validating run for this change?

* [qa: valgrind error: Leak_StillReachable calloc calloc _dl_check_map_versions](https://tracker.ceph.com/issues/65779)
* [qa: valgrind error: Leak_StillReachable malloc malloc strdup](https://tracker.ceph.com/issues/65778)

mchangir commented:
rebased and fixed conflicts with:
* `src/include/ceph_fs.h`: updated enum value for uninline data op
* `src/mds/CInode.cc`: accepted `HEAD` change with additional `bypassfreezing` parameter

mchangir commented:
rebased

vshankar commented:
@mchangir - this requires rebase (again) - unfortunately this needs to be run through qa suite once more (hopefully, the last time).

mchangir commented:
> @mchangir - this requires rebase (again) - unfortunately this needs to be run through qa suite once more (hopefully, the last time).

rebased
build in progress

lxbsz commented:
@mchangir What's the status about this PR ?  I am afraid the tracker https://tracker.ceph.com/issues/63473 is also a inline data code bug.

vshankar commented:
> @mchangir What's the status about this PR ? I am afraid the tracker https://tracker.ceph.com/issues/63473 is also a inline data code bug.

I can answer that - this was about to be merged since the it passes fs suite tests, however, a non-trivial rebase was required since other PRs got merge in the midst. So, I suggested @mchangir to rerun the tests.

lxbsz commented:
> > @mchangir What's the status about this PR ? I am afraid the tracker https://tracker.ceph.com/issues/63473 is also a inline data code bug.
> 
> I can answer that - this was about to be merged since the it passes fs suite tests, however, a non-trivial rebase was required since other PRs got merge in the midst. So, I suggested @mchangir to rerun the tests.

Cool. Get it. thanks

mchangir commented:
redoing build

mchangir commented:
[Job Set](https://pulpito.ceph.com/mchangir-2024-05-24_01:10:03-fs-wip-mchangir-mds-uninline-file-during-scrub-testing-default-smithi/) for generic fs suite

mchangir commented:
Uninline specific test jobs: [9 passed out of 12](https://pulpito.ceph.com/mchangir-2024-05-24_15:28:08-fs:functional-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/)

[1 failed job](https://pulpito.ceph.com/mchangir-2024-05-24_15:28:08-fs:functional-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/7724676) due to
```
"2024-05-25T04:38:41.105724+0000 mds.c (mds.1) 46 : cluster [WRN] evicting unresponsive client smithi105:0 (32598), after 303.942 seconds" in cluster log
```

2 dead jobs due to `hit max job timeout` of 12h
1. [dead job 1](https://pulpito.ceph.com/mchangir-2024-05-24_15:28:08-fs:functional-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/7724675)
2. [dead job 2](https://pulpito.ceph.com/mchangir-2024-05-24_15:28:08-fs:functional-wip-mchangir-mds-uninline-file-during-scrub-distro-default-smithi/7724677)

mchangir commented:
fixed python test

mchangir commented:
[rerun of fs job suite](https://pulpito.ceph.com/mchangir-2024-06-03_09:01:01-fs-wip-mchangir-testing1-uninline-main-debug-testing-default-smithi/)

mchangir commented:
> [rerun of fs job suite](https://pulpito.ceph.com/mchangir-2024-06-03_09:01:01-fs-wip-mchangir-testing1-uninline-main-debug-testing-default-smithi/)

```
 54: failed
 30: dead
156: passed
-----------
240: total
```

Run review in progress.

vshankar commented:
Spoke to @mchangir regarding this - the fs suite runs has a bunch of infra related failures and some quiesce related failures. @leonid-s-usov suggested a rebase and a rerun. I hope it's the last rebase and rerun. If we still run into infra issues and nothing related to this change itself, then I'll vote to merge this change since it has been waiting for long.

lxbsz commented:
> Spoke to @mchangir regarding this - the fs suite runs has a bunch of infra related failures and some quiesce related failures. @leonid-s-usov suggested a rebase and a rerun. I hope it's the last rebase and rerun. If we still run into infra issues and nothing related to this change itself, then I'll vote to merge this change since it has been waiting for long.

Sounds good.

mchangir commented:
build failed due to centos 9 rpms copy to chacra failed
rebased and retrying build with new branch
[Teuthology QA Tracker](https://tracker.ceph.com/issues/66375)

mchangir commented:
rebased: [QA Tracker](https://tracker.ceph.com/issues/66493)

lxbsz commented:
Another qa test failure caused by the buggy inline data https://tracker.ceph.com/issues/64707#note-23.

batrick reviewed (dismissed):
Was working on this review but got pulled into some meetings. Leaving the comments I had so far.

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -1197,3 +1354,26 @@ void ScrubStack::handle_mds_failure(mds_rank_t mds)
+  MDRequestRef mdr = in->mdcache->request_start_internal(CEPH_MDS_OP_UNINLINE_DATA);
+  std::string path;
+  in->make_path_string(path);
+  mdr->set_filepath(filepath(path));
```
done ... setting only inode in `filepath`

mchangir commented on src/mds/ScrubStack.cc:
```diff
@@ -1197,3 +1354,26 @@ void ScrubStack::handle_mds_failure(mds_rank_t mds)
+  mdr->set_filepath(filepath(path));
+  mdr->snapid = CEPH_NOSNAP;
+  mdr->no_early_reply = true;
+  mdr->internal_op_private = in;
```
`mdr->internal_op_private = in;` has been dropped

mchangir commented:
rebased and addressed Patrick's comments

mchangir commented:
Testing in Progress: [job set for suite fs](https://pulpito.ceph.com/mchangir-2024-08-02_07:51:06-fs-wip-mchangir-uninline-debug-distro-default-smithi/)

mchangir commented:
This PR is under test in [https://tracker.ceph.com/issues/67653](https://tracker.ceph.com/issues/67653).

mchangir commented:
This PR is under test in [https://tracker.ceph.com/issues/67767](https://tracker.ceph.com/issues/67767).

mchangir commented:
QA Run: https://pulpito.ceph.com/mchangir-2024-08-28_12:36:35-fs-wip-mchangir-testing-20240828.085847-main-debug-distro-default-smithi/

vshankar commented:
> QA Run: https://pulpito.ceph.com/mchangir-2024-08-28_12:36:35-fs-wip-mchangir-testing-20240828.085847-main-debug-distro-default-smithi/

There are lot many of such failures

```
2024-08-28T17:25:03.733 INFO:tasks.ceph:osd dev map: {}
2024-08-28T17:25:03.733 ERROR:teuthology.contextutil:Saw exception from nested tasks
Traceback (most recent call last):
  File "/home/teuthworker/src/git.ceph.com_teuthology_cf5021f85c4b0bf435c74a1183036b3d19af44b5/teuthology/contextutil.py", line 30, in nested
    vars.append(enter())
  File "/usr/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/home/teuthworker/src/git.ceph.com_ceph-c_d19e6118f7015f0047264006cb0c873528eb3ae6/qa/tasks/ceph.py", line 738, in cluster
    assert roles_to_devs, \
AssertionError: remote smithi067.front.sepia.ceph.com has osd roles, but no osd devices were specified!
2024-08-28T17:25:03.810 INFO:teuthology.util.scanner:summary_data or yaml_file is empty!
2024-08-28T17:25:03.849 INFO:teuthology.util.scanner:summary_data or yaml_file is empty!
```

Have you checked some recent commit that probably modified some yaml causing the osd dev map to be empty?

vshankar commented:
jenkins test make check

mchangir commented:
@vshankar 
latest run: https://pulpito.ceph.com/mchangir-2024-08-30_08:30:37-fs-wip-mchangir-testing-20240828.085847-main-debug-distro-default-smithi/

46 failed out of 313

mchangir commented:
jenkins test make check

vshankar commented:
jenkins test make check

vshankar commented:
> @vshankar latest run: https://pulpito.ceph.com/mchangir-2024-08-30_08:30:37-fs-wip-mchangir-testing-20240828.085847-main-debug-distro-default-smithi/
> 
> 46 failed out of 313

This run looks good. Do you have a run log posted in the wiki? If nothing is standing out, then this change is good to merge.

mchangir commented:
> > @vshankar latest run: https://pulpito.ceph.com/mchangir-2024-08-30_08:30:37-fs-wip-mchangir-testing-20240828.085847-main-debug-distro-default-smithi/
> > 46 failed out of 313
> 
> This run looks good. Do you have a run log posted in the wiki? If nothing is standing out, then this change is good to merge.

Run log available [here](https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-mchangir-testing-20240828085847-main-debug).

mchangir commented:
jenkins test make check

mchangir commented:
jenkins test make check arm64

vshankar commented:
> Run log available [here](https://tracker.ceph.com/projects/cephfs/wiki/Main#wip-mchangir-testing-20240828085847-main-debug).

Looks good. Nice work @mchangir

rishabh-d-dave commented:
@mchangir Can't merge this PR, proceeding to merge on his behalf.

vshankar commented:
> @mchangir Can't merge this PR, proceeding to merge on his behalf.

The tracker status is still unchanged btw.

mchangir commented:
> > @mchangir Can't merge this PR, proceeding to merge on his behalf.
> 
> The tracker status is still unchanged btw.

Tracker status updated to `Resolved`
